java-squid/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -172,13 +179,137 @@ ProgramState put(Symbol symbol, SymbolicValue value) {
     if (symbol.isUnknown()) {
       return this;
     }
-    PMap<Symbol, SymbolicValue> newValues = values.put(symbol, value);
-    if (newValues != values) {
-      return new ProgramState(newValues, constraints, visitedPoints, stack);
+    SymbolicValue oldValue = values.get(symbol);
+    if (oldValue == null || oldValue != value) {
+      PMap<SymbolicValue, Integer> newReferences = references;
+      if (oldValue != null) {
+        newReferences = decreaseReference(newReferences, oldValue);
+      }
+      newReferences = increaseReference(newReferences, value);
+      PMap<Symbol, SymbolicValue> newValues = values.put(symbol, value);
+      return new ProgramState(newValues, newReferences, constraints, visitedPoints, stack);
+    }
+    return this;
+  }
+
+  private static PMap<SymbolicValue, Integer> decreaseReference(PMap<SymbolicValue, Integer> givenReferences, SymbolicValue sv) {
+    Integer value = givenReferences.get(sv);
+    Preconditions.checkNotNull(value);
+    return givenReferences.put(sv, value - 1);
+  }
+
+  private static PMap<SymbolicValue, Integer> increaseReference(PMap<SymbolicValue, Integer> givenReferences, SymbolicValue sv) {
+    Integer value = givenReferences.get(sv);
+    if (value == null) {
+      return givenReferences.put(sv, 1);
+    } else {
+      return givenReferences.put(sv, value + 1);
+    }
+  }
+
+  ProgramState remove(Symbol symbol) {
+    if (symbol.isUnknown()) {
+      return this;
+    }
+    SymbolicValue symbolicValue = values.get(symbol);
+    if (symbolicValue != null) {
+      PMap<Symbol, SymbolicValue> newValues = values.remove(symbol);
+      PMap<SymbolicValue, Integer> newReferences = decreaseReference(references, symbolicValue);
+      PMap<SymbolicValue, Object> newConstraints = constraints;
+      if (newReferences.get(symbolicValue) == 0 && isDisposable(symbolicValue, newConstraints.get(symbolicValue)) && !inStack(stack, symbolicValue)) {
+        newConstraints = newConstraints.remove(symbolicValue);
+        newReferences = newReferences.remove(symbolicValue);
+      }
+      return new ProgramState(newValues, newReferences, newConstraints, visitedPoints, stack);
     }
     return this;
   }
 
+  private static boolean isDisposable(SymbolicValue symbolicValue, @Nullable Object constraint) {
+    return SymbolicValue.isDisposable(symbolicValue) && (constraint == null || !(constraint instanceof ObjectConstraint) || ((ObjectConstraint) constraint).isDisposable());
+  }
+
+  private static boolean inStack(Deque<SymbolicValue> stack, SymbolicValue symbolicValue) {
+    for (SymbolicValue value : stack) {
+      if (symbolicValue == value) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)